telemetry/ttl.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+const (
+	// selectDeletedRowsOneDaySQL selects the deleted rows for each table of last day
+	selectDeletedRowsOneDaySQL = `SELECT parent_table_id, CAST(SUM(deleted_rows) AS SIGNED)
+		FROM
+		    mysql.tidb_ttl_job_history
+		WHERE
+		    create_time >= CURDATE() - INTERVAL 7 DAY
+			AND finish_time >= CURDATE() - INTERVAL 1 DAY
+			AND finish_time < CURDATE()
+		GROUP BY parent_table_id;`
+	// selectDelaySQL selects the deletion delay in minute for each table at the end of last day
+	selectDelaySQL = `SELECT
+    		parent_table_id, TIMESTAMPDIFF(MINUTE, MIN(tm), CURDATE()) AS ttl_minutes
+		FROM
+			(
+				SELECT
+					table_id,
+					parent_table_id,
+					MAX(ttl_expire) AS tm
+				FROM
+					mysql.tidb_ttl_job_history
+				WHERE
+					create_time > CURDATE() - INTERVAL 7 DAY
+					AND finish_time < CURDATE()
+					AND status = 'finished'
+					AND JSON_VALID(summary_text)
+					AND summary_text ->> "$.scan_task_err" IS NULL
+				GROUP BY
+					table_id, parent_table_id
+			) t
+		GROUP BY parent_table_id;`
+)
+
+type ttlHistItem struct {
+	LessThan    *int64 `json:"less_than,omitempty"`
```

Comment:
Why using a pointer here?

Reply:
Because we should marshal `ttlHistItem` to json string. If we do not use a pointer, `LessThan` we be serialized to "0" if not set.